graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -185,37 +199,41 @@ class RenameSchemaTypesVisitor(Visitor):
             "UnionTypeDefinitionNode",
         }
     )
-
-    def __init__(self, renamings, query_type, scalar_types):
+    type_hint_rename_types = Union[
```

Comment:
this looks weird since we just defined `rename_types` above, but so far it's the only way I've been able to make the type-hints work out. [See here](https://github.com/kensho-technologies/graphql-compiler/pull/834#discussion_r436979070)

Reply:
One alternative I can think of is:
```
from typing import get_args
rename_types = {cls.__name__ for cls in get_args(type_hint_rename_types)}
```

Not sure if that's any better

Reply:
This is better than the `get_args()` alternative because this is resolvable without running the code. That means `mypy` will be able to check it, unlike @bojanserafimov's offered alternative — remember that `mypy` reads the Python AST but doesn't actually execute the code.

Reply:
I agree that it's better, just not for that reason. Mypy can still infer the type of `rename_types`, since the return type of `__name__` is `str`.

Reply:
Fair point.

@LWprogramming btw, let's avoid the `cast()` call that you linked to, and instead explicitly assert that the type of `node` is indeed one of the types in the union:
```
if not isinstance(node, <union types here>):
    raise AssertionError(<appropriate error message>)
# Here, mypy knows that if this line is reached, the type of "node" must have been appropriate, and no cast is necessary.
```

Remember that `cast()` is implicitly a "mypy, trust me, I know what I'm doing" statement. It is almost always better to use an explicit assertion rather than a mypy suppression like this.

Reply:
I tried a bunch of different ways to implement this, but ran into problems with deciding which properties I wanted to prioritize. Here are the ones that I'm thinking about:

* define the rename-types with as little repetition as possible because they basically refer to the same idea

* ensure that it's quick to check if a node is in `noop_types` or `rename_types`. If, for instance, they weren't `frozenset`s but rather lists instead, there might be performance issues with manipulating very large schemas because we'd need to check many node types for each node in the schema AST.

* type-hint `_rename_name_and_add_to_record` so that `mypy` knows the necessary attributes exist (i.e. `node` has  `name` attribute)

* avoid needing to call `cast` by raising an assertion error if the node is the wrong type

The problem is that avoiding `cast` while being able to type-hint properly seems to require even more repetition of the same idea in code. I've tried several different ways to specify `<union types here>` so that we can write the assertion: passing in the `type_hint_rename_types` Union, passing in a set of the classes involved in `rename_types` (not the class names but the classes themselves), and passing in a tuple of the classes involved. This is the only thing that worked:

```
rename_types_classes = (
    EnumTypeDefinitionNode,
    InterfaceTypeDefinitionNode,
    NamedTypeNode,
    ObjectTypeDefinitionNode,
    UnionTypeDefinitionNode,
)
...
elif node_type in self.rename_types:
    # Rename node, put name pair into record
    if not isinstance(node, self.rename_types_classes):
        raise AssertionError
```

This didn't:
```
rename_types_classes = tuple(rename_types)
```

And this didn't either:
```
if not isinstance(node, tuple(self.rename_types)):
```

In other words, the only way I've found that satisfies `mypy` while avoiding `cast` involves making another representation of the same concept.

I could attempt to reduce this duplication by using Bojan's suggestion to dynamically compute `rename_types`, but that doesn't solve the problem of needing

* a `Union` for the type hints

* a tuple to avoid `cast()`

* a set to make membership checking quick

Reply:
To be clear, I'm not opposed to using `cast` if it's accompanied by a dynamically-checked assertion of the property that `cast` is used to assure `mypy` of. In that case, `cast` is simply reiterating to `mypy` a fact that if `mypy` may in the future be able to deduce from the runtime check via more sophisticated code analysis.

My concern was simply around using `cast` in a situation where we "know" a certain fact to be "true" but aren't actually stating our certainty under penalty of raised `AssertionError`.

Reply:
In this case it might be fine to not have the assertion, since the [line of code](https://github.com/kensho-technologies/graphql-compiler/pull/847/files#diff-3b859f46e97b0af834db0b3b7de08658R286) where we use `cast` is inside a conditional saying that the node type is in `rename_types`. In a sense, the conditional functions as the assertion because we won't enter the path unless it's a type to rename. We wouldn't need to assert that `node_type` is in `rename_types` because the condition says exactly that already.

Unfortunately, this does depend on `rename_types` and `type_hint_rename_types` changing in parallel, but `mypy` complains about types at runtime (like how we described earlier in this thread) when I try to avoid the duplication by dynamically defining `rename_types`.

There might be some way to guarantee that `rename_types` and `type_hint_rename_types` have the same information-- maybe something like [this](https://stackoverflow.com/questions/45957615/check-a-variable-against-union-type-at-runtime-in-python-3-6), although it does look pretty hacky. If we do this check in `__init__`, we should be good to go, right?

Reply:
I think `mypy` should at least in part enforce that those two change together, otherwise there would be a type mismatch, no?

Dynamic check at module level is probably okay too, if you feel that's necessary.

If I may nit a bit, `type_hint_rename_types` doesn't sound like the name of a type alias though :) How about `RenameTypes` instead, or something like that?

Reply:
Unfortunately `mypy` doesn't enforce this automatically.

[Right now](https://github.com/python/mypy/issues/5354) it effectively doesn't allow us to do any introspection for `Unions`, but its maintainers did merge a PR [very recently](https://github.com/python/mypy/pull/8779) that lets us avoid the error `The type alias to Union is invalid in runtime context`. Unfortunately it's more recent than the latest `mypy` release so for now I'll tell `mypy` to ignore the dynamic check.

Reply:
Unfortunate... just discovered `get_args` is only available in Python 3.8, not 3.6 or 3.7

Reply:
Can you still use `RenameTypes.__args__`? Gives me the same result so far, but not sure if the spec is the same.

Reply:
That seems to work in [3.8](https://github.com/python/cpython/blob/3.8/Lib/typing.py#L1304) and [3.7](https://github.com/python/cpython/blob/3.7/Lib/typing.py#L618) (although not documented as far as I can tell), but the code for [3.6](https://github.com/python/cpython/blob/3.6/Lib/typing.py#L686) seems significantly different and I haven't been able to tell if this would work.

Reply:
Let's just define them twice and add a comment to keep them the same. While it's important to deduplicate where possible and makes sense, in this case I think we've put in more than a reasonable amount of work to try to make that happen.

In the end, we have to make sure the cure isn't worse than the disease, and special-casing by Python version / implementing our own `get_args` / similar options seem much worse than just a tiny bit of duplication in a place that we hopefully won't need to change anytime soon — the GraphQL AST is unlikely to get new AST nodes anytime soon.